fendermint/testing/contract-test/tests/upgrades.rs

```diff
@@ -0,0 +1,300 @@
+// Copyright 2022-2024 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+mod staking;
+
+use anyhow::{Context, Result};
+use cid::Cid;
+use std::future::Future;
+use std::sync::Arc;
+use tendermint_rpc::HttpClient;
+
+use fvm::engine::MultiEngine;
+use fvm_shared::bigint::Zero;
+use fvm_shared::clock::ChainEpoch;
+use fvm_shared::econ::TokenAmount;
+use fvm_shared::version::NetworkVersion;
+
+use fendermint_vm_actor_interface::system;
+use fendermint_vm_core::Timestamp;
+use fendermint_vm_genesis::{Genesis, PermissionMode};
+use fendermint_vm_interpreter::fvm::bundle::{bundle_path, custom_actors_bundle_path};
+use fendermint_vm_interpreter::fvm::state::{
+    FvmExecState, FvmGenesisState, FvmStateParams, FvmUpdatableParams,
+};
+use fendermint_vm_interpreter::fvm::store::memory::MemoryBlockstore;
+use fendermint_vm_interpreter::fvm::upgrade_scheduler::UpgradeScheduler;
+use fendermint_vm_interpreter::fvm::{FvmApplyRet, FvmGenesisOutput, FvmMessage, PowerUpdates};
+use fendermint_vm_interpreter::GenesisInterpreter;
+use fendermint_vm_interpreter::{
+    fvm::{bundle::contracts_path, upgrade_scheduler::Upgrade, FvmMessageInterpreter},
+    ExecInterpreter,
+};
+
+#[derive(Clone)]
+struct Tester<I> {
+    interpreter: Arc<I>,
+    state_store: Arc<MemoryBlockstore>,
+    multi_engine: Arc<MultiEngine>,
+    exec_state: Arc<tokio::sync::Mutex<Option<FvmExecState<MemoryBlockstore>>>>,
+    state_params: FvmStateParams,
+}
+
+impl<I> Tester<I>
```

Comment:
Great job extracting this from the `App`. For a moment I was thinking it could be a nice refactor, to disentangle the state management from all the other stuff that goes on there. It's not clear how much can be done, as the method parameters are slightly different there, even though the overall shape is the same. In any case it's a great reference, and I think you could move this tester to live under `testing/contract-test/src/lib.rs`, as an interpreter driver for FVM messages. It allows block boundaries, which hasn't been done so far, but would be useful in other tests I'm sure.